crates/red_knot_test/src/diagnostic.rs

```diff
@@ -0,0 +1,167 @@
+//! Sort and group diagnostics by line number, so they can be correlated with assertions.
+//!
+//! We don't assume that we will get the diagnostics in source order.
+
+use ruff_source_file::{LineIndex, OneIndexed};
+use ruff_text_size::Ranged;
+use std::ops::{Deref, Range};
+
+#[derive(Debug)]
+pub(crate) struct SortedDiagnostics<T> {
+    diagnostics: Vec<T>,
+    lines: Vec<LineDiagnosticRange>,
+}
+
+impl<T> SortedDiagnostics<T>
+where
+    T: Ranged + Clone,
+{
+    pub(crate) fn new(diagnostics: impl IntoIterator<Item = T>, line_index: &LineIndex) -> Self {
+        let mut diagnostics: Vec<_> = diagnostics
+            .into_iter()
+            .map(|diagnostic| DiagnosticWithLine {
+                line_number: line_index.line_index(diagnostic.start()),
+                diagnostic,
+            })
+            .collect();
+        diagnostics.sort_by_key(|diagnostic_with_line| diagnostic_with_line.line_number);
```

Comment:
Is a stable sort important here? No tests fail if I do this:

```suggestion
        diagnostics.sort_unstable_by_key(|diagnostic_with_line| diagnostic_with_line.line_number);
```

Reply:
Nope, good call; stable sort is not important. We attach no meaning to the order we receive the diagnostics in.